_posts/2024-02-06-new-ship-pathfinder.md

```diff
@@ -0,0 +1,93 @@
+In my opinion, it simply makes ships no fun to use.
+I often found myself terraforming narrow stretches of land into the ocean in order to create railways to access oilfields.
+Yes, I am "that guy".
+
+Having messed around quite a bit with A* pathfinders and derivatives, I always felt there had to be a better way to achieve this.
+I played around with several ideas, but most of them didn't scale well and didn't make it off the drawing board.
+Getting things to work smoothly and reliably on a 4096x4096 map turns out to be quite a challenge.
+But as they say, "if it was easy, it would have been done already".
+
+## Why is it so difficult?
+
+So why is it so hard for ships to find a path?
+If trains can do it so should ships, right?
+The main problem is that open water creates a ton of symmetric paths which are all "equally good".
+The pathfinder must explore all of these permutations in order to find the best path. 
+
+Trains don't have this issue because they are restricted by tracks and junctions.
+Potential paths branch off far less often, even in complicated networks.
+The same is true for road vehicles and trams.
+For ships however, the amount of path options grows exponentially as the distance increases.
+
+The pathfinder eventually hits the node limit and has to throw in the towel.
+Raising the node limit means gets you a bit further, but the impact on game speed becomes more noticable.
+Adding more ships further exacerbates the problem, since having twice the number of ships means twice the number of pathfinder calls.
+And since larger maps often means more players and large travel distances, one can see that the problem only gets worse as the scale increases.
+
+## The solution
+
+The solution I went for is a two-tiered pathfinding system, somewhat similar to the well-documented HPA* algorithm.
+First, we try to find a coarse, high-level path.
+If such a path is found, we choose an intermediate destination not too far from the ship.
+We use the low-level pathfinder to plan a path to that intermediate destination.
+Once we reach it, we run the high-level pathfinder again, and the process repeats. 
+
+Think of it like taking a long trip with the car: instead of planning the entire journey in meticulous detail, you just focus on how to get to the next city.
+Ironically, this is similar to how buoys work: they also provide a sequence of intermediate destinations.
+But this time it requires no user effort.
+
+To enable the high-level path search, we divide the map into 16x16 tile regions.
+Within each region, we identify separate patches of water.
+Each tile gets a label, and all tiles with the same label belong to the same patch, i.e. they are interconnected.
+We also identify whether a ship can travel across a region edge into any adjacent regions.
+Aqueducts crossing into other regions are also taken into account.
```

Comment:
Does this include aqueducts that start in one region, completely skip multiple regions before ending in another, non-adjacent region?

Reply:
Yes, it can either be to adjacent regions or with regions in between, there is no limit. I chose not to go too much into detail about aqueducts, just that "they work". Getting the pathfinder to work with aqueducts was a nightmare, but I don't think I'll do anyone a favor by ranting about it in this blog post ;)